x/leverage/keeper/inspector.go

```diff
@@ -0,0 +1,363 @@
+			specificCollateralValue, _ = k.VisibleTokenValue(ctx, sdk.NewCoins(specificCollateral), types.PriceModeSpot)
+		}
+
+		summary := types.BorrowerSummary{
+			Address:                 addr.String(),
+			SuppliedValue:           suppliedValue,
+			CollateralValue:         collateralValue,
+			BorrowedValue:           borrowedValue,
+			BorrowLimit:             borrowLimit,
+			LiquidationThreshold:    liquidationThreshold,
+			SpecificCollateralValue: specificBorrowedValue,
+			SpecificBorrowValue:     specificCollateralValue,
+		}
+		borrowers = append(borrowers, &summary)
+		return nil
+	}
+
+	// collect all borrower summaries (unsorted)
+	_ = k.iterate(ctx, prefix, iterator)
+	return borrowers
+}
+
+// riskData
+func (k Keeper) riskData(ctx sdk.Context) []types.AccountDecBalances {
+	tokens := k.GetAllRegisteredTokens(ctx)
+	symbols := map[string]string{}
+	exponents := map[string]uint32{}
+	exchangeRates := map[string]sdk.Dec{}
+	for _, t := range tokens {
+		symbols[t.BaseDenom] = t.SymbolDenom
+		exponents[t.BaseDenom] = t.Exponent
+		exchangeRates[t.BaseDenom] = k.DeriveExchangeRate(ctx, t.BaseDenom)
+	}
+
+	// which addresses have already been checked
+	accounts := []types.AccountDecBalances{}
+	checkedAddrs := map[string]interface{}{}
+
+	prefix := types.KeyPrefixAdjustedBorrow
+	iterator := func(key, _ []byte) error {
+		// get borrower address from key
+		addr := types.AddressFromKey(key, prefix)
+
+		// if the address is already checked, do not check again
+		if _, ok := checkedAddrs[addr.String()]; ok {
+			return nil
+		}
+		checkedAddrs[addr.String()] = struct{}{}
+
+		borrowed := k.GetBorrowerBorrows(ctx, addr)
+		supplied, _ := k.GetAllSupplied(ctx, addr)
+		collateral := k.GetBorrowerCollateral(ctx, addr)
+
+		symbolBorrowed := symbolDecCoins(borrowed, symbols, exponents, exchangeRates)
+		symbolSupplied := symbolDecCoins(supplied, symbols, exponents, exchangeRates)
+		symbolCollateral := symbolDecCoins(collateral, symbols, exponents, exchangeRates)
+
+		accounts = append(accounts, types.AccountDecBalances{
+			Supplied:   symbolSupplied,
+			Collateral: symbolCollateral,
+			Borrowed:   symbolBorrowed,
+		})
+
+		return nil
+	}
+
+	// collect all account symbol balances (unsorted)
+	_ = k.iterate(ctx, prefix, iterator)
+	return accounts
+}
+
+// symbolDecCoins converts an sdk.Coins containing base tokens or uTokens into an sdk.DecCoins containing symbol denom
+// base tokens. for example, 1000u/uumee becomes 0.0015UMEE at an exponent of 6 and uToken exchange rate of 1.5
+func symbolDecCoins(
+	coins sdk.Coins,
+	symbols map[string]string,
+	exponents map[string]uint32,
+	exchangeRates map[string]sdk.Dec,
+) sdk.DecCoins {
+	symbolCoins := sdk.NewDecCoins()
+
+	for _, c := range coins {
+		if _, ok := symbols[c.Denom]; !ok {
+			// unregistered tokens cannot be converted, but can be returned as base denom
+			symbolCoins = symbolCoins.Add(sdk.NewDecCoinFromDec(c.Denom, sdk.NewDecFromInt(c.Amount)))
```

Comment:
why we don't apply exchange rate here?

Reply:
an unregistered denom will not have a uToken or a valid utoken exchange rate

(in practice, unregistered tokens shouldn't appear here at all, but if they do I want them to be displayed as-is rather than causing an error)